update docs for release v0.4.0 (#778)
ray-project/kuberay: apiserver/README.md

# KubeRay APIServer

The KubeRay APIServer provides gRPC and HTTP APIs to manage KubeRay resources.

**Note**

    The KubeRay APIServer is an optional component. It provides a layer of simplified
    configuration for KubeRay resources. The KubeRay API server is used internally
    by some organizations to back user interfaces for KubeRay resource management.

    The KubeRay APIServer is community-managed and is not officially endorsed by the
    Ray maintainers. At this time, the only officially supported methods for
    managing KubeRay resources are

    - Direct management of KubeRay custom resources via kubectl, kustomize, and Kubernetes language clients.
    - Helm charts.

    KubeRay APIServer maintainer contacts (GitHub handles):
    @Jeffwan @scarlet25151

## Usage

> You can just install the KubeRay APIServer within the same kubernetes cluster by using the [helm chart](https://github.com/ray-project/kuberay/tree/master/helm-chart/kuberay-apiserver) or just use [kustomize](https://github.com/ray-project/kuberay/tree/master/apiserver/deploy/base)

Reviewer: ```suggestion
You can install the KubeRay APIServer by using the [helm chart](https://github.com/ray-project/kuberay/tree/master/helm-chart/kuberay-apiserver) or [kustomize](https://github.com/ray-project/kuberay/tree/master/apiserver/deploy/base)
```

After the deployment we may use the `{{baseUrl}}` to access the 

- (default) for nodeport access, we provide the default http port `31888` for connection and you can connect it using.

- for ingress access, you will need to create your own ingress 

The requests parameters detail can be seen in [KubeRay swagger](https://github.com/ray-project/kuberay/tree/master/proto/swagger), here we only present some basic example:

## Setup end-to-end test

0. (optional) you may use your local kind cluster or minikube

```bash

> kind create cluster --name ray-test

Reviewer 2: If users want to use the NodePort service in the KinD cluster, the KinD config needs to set `extraPortMappings` properly. Maybe we can reuse existing KinD configs or provide a new config.

Author: yes, I pushed a modification here with complete kind config in the readme.md workflow.

```

1. Deploy the KubeRay APIServer within the same cluster of KubeRay operator 

```bash
helm -n ray-system install kuberay-apiserver kuberay/helm-chart/kuberay-apiserver
```

> 1. you can access by your nodeport

Reviewer 2: 2 ?

Author: Fix typos

```
curl localhost:31888
{"code":5, "message":"Not Found"}
```

> 3. you can create `RayCluster` or `RayJobs` or  `RayService` by dialing the endpoints

Reviewer 2: Could you explain what's the goal of this `curl` command` and add a link to APIServer's API document?

Author: here I think there is brief guide of using apiserver to create resource in kuberentes cluster and for detail definition, we provide the [swagger support](https://ray-project.github.io/kuberay/components/apiserver/#swagger-support).

Reviewer 2: How about "You can create `RayCluster`, ' RayJobs` or  `RayService` by dialing the endpoints. See [swagger support](https://ray-project.github.io/kuberay/components/apiserver/#swagger-support) for more details."?

> 3. you can just create `RayCluster` or `RayJobs` or  `RayService` by just dials the endpoints

Reviewer: ```suggestion
3. you can create `RayCluster` or `RayJobs` or  `RayService` by dialing the endpoints
```

```

> curl -XPOST 'localhost:31888/apis/v1alpha2/namespaces/ray-system/services' \

Reviewer 2: `-X POST`?

Author: `curl -XPOST` is valid and will works, and for better reading format I modified to `-X POST`

--header 'Content-Type: application/json' \
--data '{
  "name": "user-test-1",
  "namespace": "ray-system",
  "user": "user",
  "serveDeploymentGraphSpec": {
      "importPath": "fruit.deployment_graph",
      "runtimeEnv": "working_dir: \"https://github.com/ray-project/test_dag/archive/c620251044717ace0a4c19d766d43c5099af8a77.zip\"\n",
      "serveConfigs": [
      {
        "deploymentName": "OrangeStand",
        "replicas": 1,
        "userConfig": "price: 2",
        "actorOptions": {
          "cpusPerActor": 0.1
        }
      },
      {
        "deploymentName": "PearStand",
        "replicas": 1,
        "userConfig": "price: 1",
        "actorOptions": {
          "cpusPerActor": 0.1
        }
      },
      {
        "deploymentName": "FruitMarket",
        "replicas": 1,
        "actorOptions": {
          "cpusPerActor": 0.1
        }
      },{
        "deploymentName": "DAGDriver",
        "replicas": 1,
        "routePrefix": "/",
        "actorOptions": {
          "cpusPerActor": 0.1
        }
      }]
  },
  "clusterSpec": {
    "headGroupSpec": {
      "computeTemplate": "default-template",
      "image": "rayproject/ray:2.1.0",
      "serviceType": "NodePort",

>       "rayStartParams": {

Reviewer 2: @DmitriGekhtman Is there any redundant field here?

Reviewer: "port" and "node-ip-address" are unnecessary

            "port": "6379",
            "node-ip-address": "$MY_POD_IP",
            "dashboard-host": "0.0.0.0",
            "metrics-export-port": "8080"
        },
       "volumes": [] 
    },
    "workerGroupSpec": [
      {
        "groupName": "small-wg",
        "computeTemplate": "default-template",
        "image": "rayproject/ray:2.1.0",
        "replicas": 1,
        "minReplicas": 0,
        "maxReplicas": 5,
        "rayStartParams": {
                "node-ip-address": "$MY_POD_IP"
            }
      }
    ]
  }
}'
```

> then you can have the resource running in kubernetes cluster.

Reviewer: ```suggestion
The Ray resource will then be created in your Kubernetes cluster.
```
